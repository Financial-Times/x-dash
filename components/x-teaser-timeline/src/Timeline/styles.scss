@import '@financial-times/o-typography/main';

.container {
	display: flex;
	flex-direction: column;
}
