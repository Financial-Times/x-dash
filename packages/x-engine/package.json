{
  "name": "@financial-times/x-engine",
  "version": "1.0.0-1",
  "description": "A consolidation library to render x- components with any compatible runtime.",
  "main": "src/server.js",
  "browser": "src/client.js",
  "keywords": [],
  "author": "",
  "license": "ISC",
  "dependencies": {
    "assign-deep": "^0.4.7"
  },
  "engines": {
    "node": ">= 6.0.0"
  },
  "publishConfig": {
    "access": "public"
  }
}
