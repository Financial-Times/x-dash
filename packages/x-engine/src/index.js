const path = require('path');
const deepGet = require('./concerns/deep-get');

const cwd = process.cwd();

// 1. try to load the application's package manifesr
const manifest = path.join(cwd, 'package.json');

let pkg;

try {
	pkg = require(manifest);
} catch(error) {
	throw new Error(`Could not load package.json from ${cwd}`);
}

// 2. if we have the manifest then find the engine configuration
const runtime = deepGet(pkg, 'x-dash.engine.server');

if (typeof runtime !== 'string') {
	throw new Error('x-engine requires a server runtime engine to be specified');
}

// 3. if this module is a linked dependency then resolve required runtime to CWD
const symlink = __dirname.startsWith(cwd) === false;
const dependency = symlink ? path.join(cwd, 'node_modules', runtime) : runtime;

let engine;

try {
	engine = require(dependency);
} catch (error) {
	// TODO: A friendlier error message
	throw new Error(`Failed to load ${runtime} from ${dependency}`);
}

// 4. if we've loaded the runtime then find it's create element factory function
const factory = typeof engine === 'function' ? engine : (engine.createElement || engine.h);

module.exports = factory;
