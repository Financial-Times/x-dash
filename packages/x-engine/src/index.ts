import { SFC, Attributes, ReactNode, ReactElement } from 'react';
import { join } from 'path';
import { readFileSync } from 'fs';
import deepGet from './concerns/deep-get';

// All components can extend a generic functional component without explicitly
// importing any React typings.
export type Component<P> = SFC<P>;

// Any engine we load should expose have a generic factory interface
export type Factory = <P>(
	type: Component<P> | string,
	props?: Attributes & P | null,
	...children: ReactNode[]
) => ReactElement<P> | string;

const manifest = join(process.cwd(), 'package.json');

const pkg = JSON.parse(readFileSync(manifest).toString());

const runtime = deepGet(pkg, 'x-dash.engine.server');

if (typeof runtime !== 'string') {
	throw new Error('x-engine requires a server runtime engine to be specified');
}

let engine;

try {
	engine = require('./engines/' + runtime) as Factory;
} catch (error) {
	// TODO: A friendlier error message
	throw new Error(`Failed to load "${runtime}"`);
}

export const h = engine;
