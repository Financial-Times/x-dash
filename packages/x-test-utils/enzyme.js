const Enzyme = require('enzyme')
const Adapter = require('enzyme-adapter-react-16')
require('jest-enzyme')

Enzyme.configure({ adapter: new Adapter() })

module.exports = Enzyme
