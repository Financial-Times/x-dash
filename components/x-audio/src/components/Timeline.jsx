import { h, Component } from '@financial-times/x-engine';
import classNameMap from './classnames-helper';
import Loading from './Loading';
import ErrorMessage from './ErrorMessage';
import { ScrubBar } from './ScrubBar'
import formatTime from './format-seconds-to-hmmss';
import { TimeRemaining } from './TimeRemaining';

export class Timeline extends Component {
	constructor(props) {
		super(props);
		this.updateCurrentTime = this.updateCurrentTime.bind(this);
		this.startScrub = this.startScrub.bind(this);
		this.finishScrub = this.finishScrub.bind(this);
		this.state = {
			currentTime: 0,
			scrubbing: false
		}
	}

	componentDidMount() {
		this.updateCurrentTime(this.props.currentTime);
	}

	componentDidUpdate (prevProps) {
		if (this.state.scrubbing && prevProps.seeking && !this.props.seeking) {
			this.setState({ scrubbing: false });
			this.props.onScrub({ isScrubbing: false });
		}
	}

	updateCurrentTime(currentTime) {
		this.setState({ currentTime })
	}

	startScrub() {
		this.setState({ scrubbing: true })
		this.props.onScrub({ isScrubbing: true });
	}

	finishScrub() {
		this.props.updateCurrentTime({ currentTime: this.state.currentTime })
	}

	render() {
		const { duration, loading, error } = this.props;
		const currentTime = this.state.scrubbing ? this.state.currentTime : this.props.currentTime;
		return (
			<div className={classNameMap('audio-player__timeline')}>
				<ScrubBar
					onChange={this.updateCurrentTime}
					onStartScrub={this.startScrub}
					onFinishScrub={this.finishScrub}
					playheadPosition={currentTime}
					duration={duration}
				/>
				{ error && <ErrorMessage />}
				{!error && loading && <Loading expanded />}
				{!error && !loading && <div className={classNameMap('audio-player__info__current-time')}>{formatTime(currentTime)}</div>}
				<TimeRemaining currentTime={currentTime} duration={duration} expanded />
			</div>
		)
	}
}
