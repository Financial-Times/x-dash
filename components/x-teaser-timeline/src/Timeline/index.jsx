import { h } from '@financial-times/x-engine';
import { ftTime } from '@financial-times/ft-date-format';
import TimelineGroup from '../TimelineGroup';

import styles from './styles.scss';

const isSameDay = (date1, date2) =>
	date1.getDate() == date2.getDate() &&
	date1.getMonth() == date2.getMonth() &&
	date1.getFullYear() == date2.getFullYear();

const isToday = someDate => isSameDay(new Date(), someDate);

const TeaserTimeline = ({ teasers = [] }) => {
	const teaserGroups = teasers
		.sort((teaser1, teaser2) => {
			return teaser2.publishTimestamp > teaser1.publishTimestamp;
		})
		.reduce((dayGroups, { publishTimestamp, Teaser }) => {
			const publishDate = new Date(publishTimestamp);
			const groupId = `${publishDate.getFullYear()}${publishDate.getMonth()}${publishDate.getDate()}`;
			const existingIndex = dayGroups.findIndex(group => group.id === groupId);

			if (existingIndex === -1) {
				dayGroups.push({
					id: groupId,
					title: isToday(publishDate) ? 'Earlier Today' : ftTime(publishDate),
					teaserComponents: [Teaser],
				});
			} else {
				dayGroups[existingIndex].teaserComponents.push(Teaser);
			}

			return dayGroups;
		}, []);

	return (
		<div className={styles.container}>
			{teaserGroups.map(group => <TimelineGroup key={group.title} {...group} />)}
		</div>
	);
};

export default TeaserTimeline;
