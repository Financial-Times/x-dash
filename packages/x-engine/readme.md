# x-engine

This module is a consolidation library to render `x-dash` components with any compatible runtime.

## Installation

This module is supported on Node 12 and is distributed on npm.

```bash
npm install -S @financial-times/x-engine
```

You'll also need to install your chosen runtime and any related dependencies. Some compatible runtimes are:

- [Hyperapp](https://github.com/hyperapp/hyperapp)<sup>\*</sup>
- [Hyperons](https://github.com/i-like-robots/hyperons)
- [Inferno](https://infernojs.org/)
- [Nerv](https://github.com/NervJS/nerv)
- [Preact](https://preactjs.com/)
- [Rax](https://alibaba.github.io/rax/)
- [React](https://reactjs.org/)
- [VDO](https://github.com/DylanPiercey/vdo)

\* Usage of Hyperapp depends on a small modification to higher-order components to accept `children` as a second argument rather than receiving them appended to `props`.

## Configuration

To start you must specify your runtime configuration within `package.json`. This instructs `x-engine` which modules to load for your environment. You may specify different runtimes for server and browser rendering depending on your needs.

This module includes several presets for popular tools and frameworks including React, Preact, and Hyperons. For example to use Hyperons on the server and Preact in the browser you may use the name of their preset:

```json
{
	"x-dash": {
		"engine": {
			"server": "hyperons",
			"browser": "preact"
		}
	}
}
```

But if your chosen tool does not have a preset then the configuration for it can be provided with the expanded configuration format. The example below shows how to load the VDO module and use its `createElement` factory function<sup>\*</sup>:

```json
{
	"x-dash": {
		"engine": {
			"server": {
				"runtime": "vdo",
				"factory": "createElement"
			}
		}
	}
}
```

\* A JSX factory function is a variadic function<sup>†</sup> with the signature `fn(element, properties, ...children)`, examples include `React.createElement` and `Preact.h`. See the [FAQ section](#faq) for more information.

† Variadic means that the function accepts a variable number of arguments. The `...` before the last arguments name is a rest parameter, meaning it will collect "the rest" of the arguments.

## Rendering

### Server-side

If your chosen runtime factory returns a string then you can pass properties to the component and immediately use the returned value:

```js
const { Teaser } = require('@financial-times/x-teaser');

app.get('/teaser', (request, response) => {
	const properties = { … };
	response.send(Teaser(properties));
});
```

But if your factory method returns a framework-specific intermediary code then you'll need to load the specified render method to render the node into a string or stream:

```js
const { render } = require('@financial-times/x-engine');
const { Teaser } = require('@financial-times/x-teaser');

app.get('/teaser', (request, response) => {
	const properties = { … };
	const nodes = Teaser(properties);
	response.send(render(nodes));
});
```

### Client-side

To use components on the client-side you will first need to add the `x-engine` plugin to your Webpack configuration file. Under the hood this uses [`DefinePlugin`](https://webpack.js.org/plugins/define-plugin/) to wire up your chosen runtime.

```js
// webpack.config.js
const xEngine = require('@financial-times/x-engine/src/webpack');

module.exports = {
	plugins: [
		xEngine()
	]
};
```

You can then install and use `x-` components in your client-side code:

```jsx
import { h } from '@financial-times/x-engine';
import { Teaser } from '@financial-times/x-teaser';

export default const TeaserList = (props) => (
	<ul class="TeaserList">
		{props.items.map((item) => (
			<li className="TeaserList-Item">
				<Teaser {...item} layout="small" showImage={true} />
			</li>
		))}
	</ul>
);
```

#### Client-side through n-ui
In order to configure n-ui to compile x-dash based `.jsx` files you will also need to add the following to `n-ui-build.config.js`.

```js
// n-ui-build.config.js
const xEngine = require('@financial-times/x-engine/src/webpack');

module.exports = {
	plugins: [
		xEngine()
	],
	pragma: 'h'
};
```

You will also need to import the factory function from `x-engine` in each `.jsx` file.

```javascript
import { h } from '@financial-times/x-engine'
```

## FAQ

### This sounds complicated… is it a magic black box?

There is no magic. The source code for the server-side integration is less than 60 lines of unexciting code. The Webpack plugin for client-side usage is even smaller.

### What is a "factory function"?

A factory function is a variadic function with the signature `fn(element, properties, ...children)`, examples include `React.createElement` and `Preact.h`. It may return the framework's representation of a HTML node or a formatted string depending on the runtime you're using.

### Which runtime should I use?

Whichever one you want! React, Preact, Rax, and Nerv are all largely compatible with one another. If you don't want the overhead of a framework, or are rendering static HTML, then it's worth investigating the Hyperons or VDO modules.

### What about Hyperscript?

Hyperscript currently only supports passing a tag name (a string) as the first argument. This limitation means you cannot currently reference components inside other components.
