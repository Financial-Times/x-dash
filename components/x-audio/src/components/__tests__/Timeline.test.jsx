import { h } from '@financial-times/x-engine';
import { mount } from '@financial-times/x-test-utils/enzyme';

jest.mock('../classnames-helper', () => jest.fn());

const MockScrubBar = () => <div className='scrub-bar'/>
jest.mock('../ScrubBar', () => ({
	ScrubBar: props => <MockScrubBar {...props} />
}));

const MockTimeRemaining = () => <div className='time-remaining'/>
jest.mock('../TimeRemaining', () => ({
	TimeRemaining: props => <MockTimeRemaining {...props} />
}));

const MockLoading = () => <div className='loading'/>
jest.mock('../Loading', () => (props => <MockLoading {...props} />));

const MockErrorMessage = () => <div className='error-message'/>
jest.mock('../ErrorMessage', () => (props => <MockErrorMessage {...props} />));

import { Timeline } from '../Timeline';

const render = (overrides = {}) => {
	const props = {
		duration: 10,
		currentTime: 0,
		seeking: false,
		onScrub: jest.fn(),
		updateCurrentTime: jest.fn(),
		loading: false,
		error: false,
		...overrides,
	}
	return {
		component: mount(<Timeline {...props}/>),
		props
	}
}

describe('Timeline', () => {

	it('should render scrub bar', () => {
		const { component } = render();
		expect(component.find(MockScrubBar)).toExist();
	});

	it('should render loading component if loading is true', () => {
		const { component } = render({ loading: true });
		expect(component.find(MockLoading)).toExist();
	});

	it('should render error message if error is true', () => {
		const { component } = render({ error: true });
		expect(component.find(MockErrorMessage)).toExist();
	});

	it('should render current time based on the playing audio', () => {
		const { component } = render({ currentTime: 10 });
		const { currentTime } = component.find(MockTimeRemaining).props()
		expect(currentTime).toEqual(10);
	});

	describe('when scrubbing', () => {

		it('should call onScrub when scrubbing starts', () => {
			const { component, props } = render();
			const { onStartScrub } = component.find(MockScrubBar).props();
			onStartScrub();
			expect(props.onScrub).toHaveBeenCalledWith({ isScrubbing: true });

		});

		it('should call updateCurrentTime when scrubbing finishes', () => {
			const { component, props } = render();
			const { onStartScrub, onChange, onFinishScrub } = component.find(MockScrubBar).props();

			onStartScrub();
			onChange(20);
			onFinishScrub();

			expect(props.updateCurrentTime).toHaveBeenCalledWith({ currentTime: 20 });

		});

		it('should render current time based on the position of the scrub bar head', () => {
			const { component } = render();
			const { onChange, onStartScrub } = component.find(MockScrubBar).props();

			onStartScrub();
			onChange(20);

			component.update();

			const { currentTime } = component.find(MockTimeRemaining).props()
			expect(currentTime).toEqual(20);
		});

		it('should not render current time from the audio element whilst it is seeking', () => {
			const { component } = render();
			const { onStartScrub } = component.find(MockScrubBar).props();

			onStartScrub();
			component.setProps({ seeking: true, currentTime: 5 });

			component.update();

			const { currentTime } = component.find(MockTimeRemaining).props()
			expect(currentTime).toEqual(0);
		});
	});
});
