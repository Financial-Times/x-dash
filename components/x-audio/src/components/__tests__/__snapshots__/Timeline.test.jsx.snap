// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Timeline given error is false should display necessary components if loading is false 1`] = `
<div
  className="audio-player__timeline"
>
  <ScrubBar
    duration={600}
    onChange={[Function]}
    onFinishScrub={[Function]}
    onStartScrub={[Function]}
    playheadPosition={0}
  />
  <div
    className="audio-player__info__current-time"
  >
    00:00
  </div>
  <TimeRemaining
    currentTime={0}
    duration={600}
    expanded={true}
  />
</div>
`;
